src/power_grid_model/_core/power_grid_model.py

```diff
@@ -108,9 +108,18 @@ def copy(self) -> "PowerGridModel":
         new_model._all_component_count = self._all_component_count
         return new_model
 
-    def __copy__(self):
+    def __copy__(self) -> "PowerGridModel":
         return self.copy()
 
+    def __deepcopy__(self, memo: dict[int, Any]) -> "PowerGridModel":
+        # PowerGridModel.copy makes already a deepcopy
```

Comment:
Do you need to check if `id(self)` is already in the `memo`? Otherwise we do a lot of copies if this object is referenced multiple times in upper layer object?